### PR TITLE
[TD] fix unit mismatch bug

### PR DESCRIPTION
**Pull request by donovaly**
(also make global decimal option italic since it only affects new dimensions)

bug report: https://forum.freecadweb.org/viewtopic.php?f=35&t=52545

Now it works as it should:
- if unit is shown, the value is kept as is:
FreeCAD_d_TifUxicXrE
- if no unit is shown, the value must be in the document's base unit:
FreeCAD_d_XUkIF3Tzra

### CODE REVIEW

**Comment by donovaly** [comment]
Don't merge because I just noticed, that it introduces a bug with the imperial schemes when values are displayed without units. I'll improve the PR later today.

**Comment by luzpaz** [comment]
@donovaly please use the new [Draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) feature instead of adding `[dont merge]` to the title of the PR. Thank you

**Comment by donovaly** [comment]
> @donovaly please use the new [Draft pull request]

It was meant to be a normal PR. Some hours later I found an issue. And hopefully soon it will become a normal PR again.

**Comment by donovaly** [comment]
Now it is really ready. In https://forum.freecadweb.org/viewtopic.php?f=35&t=52545&p=453300#p453300 I show that it works now for also for the previously failing cases.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (chennes) [inline src/Mod/TechDraw/App/DrawViewDimension.cpp:631]:
```suggestion
            // if the unit is a metric length unit, nothing has to be converted
```

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (chennes) [inline src/Mod/TechDraw/App/DrawViewDimension.cpp:653]: Is the intent really to isolate positive `userVal`, or to eliminate `userVal == 0.0`? If you are really just avoiding the div by zero, it would be clearer to state that, e.g.
```suggestion
        if (userVal != 0.0 && abs(1 - formattedValue.toDouble() / userVal) > 0.1) {
```
(I ask because I spent a disturbingly long time trying to figure out when `userVal` would be less than zero just now... 🤦)

**Review by chennes — COMMENTED** [review]
I can't check this out right now (`gh pr checkout 4094` returns `[rejected]` -- maybe it's out-of-date with the master? I'm not sure...) so just two minor comments based on a reading of the code alone. I'll recheck the PR branch tomorrow to actually test it out.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (wwmayer) [inline src/Mod/TechDraw/App/DrawViewDimension.cpp:632]: This whole if-else block is superfluous. Base::Quantity.getUserString() has an overloaded version that gives you the scale factor for free.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (wwmayer) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:653]: You have an expression of the form: abs(1 - x/y) > 0.1
This can be easily changed to the equivalent expression: abs(y - x) > 0.1 * y
This has two advantages:
* You don't have to check whether y is 0
* You avoid the rather expensive division by y

**Review by aapo-aapo — CHANGES_REQUESTED** [review]
Otherwise I find this PR improving the current TD unit system problems, but I'd suggest amending the PR by moving the logic involving literal metric units ("nm", "mm", etc.) away from the TechDraw side by using the Unit api functions.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:632]: I don't understand. I mean, yes, I should replace
`double convertValue = Base::Quantity::parse(QString::fromLatin1("1") + UnitString).getValue();`
by
`asQuantity.getUserString(convertValue, UnitString);`
(also in DlgUnitsCalculator::valueChanged)

But the if-else must stay. Take for example the case that val = 0.050000. The unit system is standard. This gives you a factor of 0.001 but in this case the factor should be 1 since 0.05000 is already in mm. If in contrary the unit system is Imperial civil, then the desired unit is ft and we must convert 0.050000 to 0.000164

What would help is a method to get the base length unit for the currently used unit system.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:599]: '°' then you opened the file with ASCII encoding instead of UTF-8.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:653]: > This can be easily changed to the equivalent expression: abs(y - x) > 0.1 * y

Thanks.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:658]: >  If the user has set the Format Spec to %.1f, it might be by accident or by intent,

Setting to one decimal for a value < 0.1 is a mistake and we discussed in the forum that outputting "0.0" is a no-go. Such a drawing is useless and violates the rules for a technical drawing. Such things occur either if you accidentally change the unit system to something unsuitable for an existing drawing or you make the mistake of setting a global decimal number that is too low for your use case. In both cases you see your mistake and can either change the unit system or increase the number of decimals.
A typical use case are, however, real-life drawings where I only need 2 decimals (cannot be milled more precisely) and only for very special fine ones I need 3 (the 3D printer has a resolution of 25 µm). So also in the past I only changed this special dimension to have 3 decimals to not pollute the whole drawing with zeros in the dozen other dimensions.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (aapo-aapo) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:599]: I tried to open it as UTF-8, but what you said could well be true, if I made a mistake with my settings.  I mean, if the character works for everyone else but me, then mistake is probably in my end.  :-)

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (aapo-aapo) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:658]: Yes, what you said is true; and that's what I like in your PR; i.e. it fixes the real-life use case.  I have no objection here, if others accept this, too.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (chennes) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:599]: It works for me in both vim and Visual Studio.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (wwmayer) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:632]: > But the if-else must stay.

Why does it handle then µm, mm, nm, m and km but ignores things like cm, dm, pm, ...? What if in the future the core system will support some exotic units like dam (deka meter)?

You say for ImperialCivil the length is in ft but what about the other imperial systems? What is their expected unit for lengths?

IMO, depending on the value of the enum "unitSystem" you must hard-code the scale factors because only then you can guarantee that the numbers are consistent.
In your example file you have chosen values where each time the quantity uses the same scale factor and thus you won't see the possible inconsistency.

But as soon as you have numbers where different scale factors are used (e.g. one value is given in inch and another value in thou) you get wrong numbers.

**Comment by aapo-aapo** [comment]
Hi, when I checked with another document (the one I used in [https://forum.freecadweb.org/viewtopic.php?f=35&t=52621&sid=52f82b4f6233acbcb43300c934bac770](url) for other PR example), I noticed that this PR increases the number precision even if the values are not close to zero, at least in some cases.  This was probably not intended, so this should also be fixed.  See the picture, it's the same situation with current FreeCAD master without and with your PR applied.



20201120_FreeCAD_TechDraw_FixUnitMismatch-pic01

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp:632]: > Why does it handle then µm, mm, nm, m and km but ignores things like cm, dm, pm, ...? 

I think this is not necessary since getUserString() (precisely UnitsSchemaInternal::schemaTranslate() ) only outputs different units in steps of 1000, even if the base unit of the scheme is cm:
FreeCAD_8iTh0jPpbl

> You say for ImperialCivil the length is in ft but what about the other imperial systems? What is their expected unit for lengths?

Good point. I'll follow your suggestion and determine this.

**Comment by donovaly** [comment]
> Hi, when I checked this with another document (the one I used for [other case](https://forum.freecadweb.org/viewtopic.php?f=35&t=52621)), I noticed that the proposed changes here increase the number precision when using "%g" Format Spec

Thanks, I will have a look. But, doesn't mean %g not exactly what you now get? You did not specify a decimal limit.

However, I'll set the number of decimals when we have %g to keep existing older drawings as they are.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (donovaly) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp]: I just pushed an improved PR where I found a solution that doesn't need the if-else and also works for cases with different units:
FreeCAD_d_VVWyVcdFNl
FreeCAD_d_NkRiFTXwSu

**Comment by aapo-aapo** [comment]
> But, doesn't mean %g not exactly what you now get? You did not specify a decimal limit.

That is correct, but historically there has been two limits for the decimals:  (1) internal number of decimals, and (2) format specifier.  It's a bit weird, but that's how it's been.  If it'll be agreed to drop historical baggage, I'd suggest an algorithm that works this way:  (1) forget internal number of decimals, (2) use format specifier with fixed number of decimals (e.g. "%.2f" as default), and (3) from the value string, drop any zeros after decimal separator, and also drop decimal separator if it's the last character.


> However, I'll set the number of decimals when we have %g to keep existing older drawings as they are.

Indeed, it's a decision about the older drawings and existing functionality.  At 1st, I thought your PR only added precision if the printed value would be > 10 % off of the correct value, which IMO would have been good thing with the old drawings, too, but according to my test that logic does not always work for some reason.

File: src/Mod/TechDraw/App/DrawViewDimension.cpp
Comment (wwmayer) [reply src/Mod/TechDraw/App/DrawViewDimension.cpp]: > I think this is not necessary since getUserString() (precisely UnitsSchemaInternal::schemaTranslate() ) only outputs different units in steps of 1000, even if the base unit of the scheme is cm:

But are again relying on implementation details of how it works. This is always bad practice because your code fails as soon as the implementation changes.

> I just pushed an improved PR where I found a solution that doesn't need the if-else and also works for cases with different units:

Yes, this looks good and should reliably avoid inconsistencies.

> Note that asQuantity.getUserString(convertValue, UnitString); cannot be used because it delivers for e.g. val = 0.050 and Standard scheme a conversion factor 0.001 but in this case the factor must be 1. Now I remember why we have what we have in DlgUnitsCalculator::valueChanged() and that's the way it works for all values.

OK.

**Comment by wwmayer** [comment]
From my side this PR can be merged. Does everybody agree?

**Comment by donovaly** [comment]
> From my side this PR can be merged. Does everybody agree?

I am reluctant because of the issue @aapo-aapo reported: FC used %g by default in the past and it limited the number of decimals despite it used %g and not %.3g or something. With my PR %g works as it should but it will change this way change the number of decimals in older existing drawings.
So merging my PR as it drops historical baggage for the price that users might be unhappy.

**Comment by donovaly** [comment]
> If it'll be agreed to drop historical baggage, I'd suggest an algorithm that works this way: (1) forget internal number of decimals, (2) use format specifier with fixed number of decimals (e.g. "%.2f" as default), and (3) from the value string, drop any zeros after decimal separator, and also drop decimal separator if it's the last character.

(1) and (2) is already the case. For (2), we use the global number of decimals set in the preferences when the corresponding TD option is checked, otherwise "%.2f".
For (3) my the technician explained me, the number of decimals also determines the precision. Therefore the number of decimals should be uniform. Having here and there a value with a higher precision is thereby OK because this points out that the precision is necessary. But having a mixture of e.g. "2", "2.2" and "2.25" instead of "2.00", "2.20" and "2.25" seems to be uncommon/undesired.
However, let's discuss this in the forum since this is independent of this PR and I am not an expert in this field.

**Comment by donovaly** [comment]
> With my PR %g works as it should but it will change this way change the number of decimals in older existing drawings.
> So merging my PR as it drops historical baggage for the price that users might be unhappy.

Hmm, I think for a clean solution we must drop the historical baggage because when I limit "%g" to keep old drawings as the are, then "%g" for new drawings won't work as the user expects. OK, "%g" for a technical drawing is not quite useful but it would be weird that "%g" will not work as in every other area of FC (and in other programs).
What is your opinion?

**Comment by aapo-aapo** [comment]
> Hmm, I think for a clean solution we must drop the historical baggage because when I limit "%g" to keep old drawings as the are, then "%g" for new drawings won't work as the user expects.
> What is your opinion?

I'd prefer to go for the clean solution, but of course my opinion should really have only as small weight as for any other user.  Keeping old drawings intact doesn't matter much to me, personally.  And, as a programmer I like to use a specific Format Spec instead of "number of decimals", anyways.  Of course, this is bound to cause complaints from some users, so it should really be a maintainer decision.

**Comment by wwmayer** [comment]
> It'd good if @wwmayer gave his opinion as a maintainer as you asked

Since I don't use TechDraw that often I am reluctant to make a decision. IMO, this should be up to the developers that use the WB every day.

**Comment by donovaly** [comment]
> Since I don't use TechDraw that often I am reluctant to make a decision. IMO, this should be up to the developers that use the WB every day.

That would for example be me ;-). I made a lot of TD contributions already because I use TD frequently. So let's be progressive: 
I will extend this PR to limit the number of decimals in case of %g if the preferences setting is to use the global number of decimals. I think this is a fair compromise because then the user's will is respected and that way we have an easy workaround for old drawings - the user just needs to change the TD preferences.

**Comment by wwmayer** [comment]
I am going to merge PR #4087. Be aware of that this causes some minor conflicts with your branch as soon as you rebase it.

**Comment by donovaly** [comment]
The PR is ready to be merged. I made the discussed change that existing older drawings keep as they are and also fixed a further minor glitch: when the angular unit is "deg" there must be a space between the dimension and unit.

**Comment by wwmayer** [comment]
Merged.

Had to solve a merge conflict.

**Comment by luzpaz** [comment]
Adding commit for posterity: 0baabcaa4564805276d4230cdcc8838711181317

**Comment by aapo-aapo** [comment]
After this PR was merged, the number of decimals displayed changes for me when toggling "Preferences->TechDraw->Dimensions->Show Units" on/off, if the Dimension is using "%g" format spec, which is odd (and probably a bug).  Also, I got some errors in the report window, like `Warning - no numeric format in formatSpec %.2 - Dimension001`, which seems to imply that something is cutting the format specifier short.

**Comment by donovaly** [comment]
> This only happens, if I don't have "Use global decimals" set at all

Yes, we agreed on this (you gave me a thumbs up). It is the compromise.
%g does not limit the number of decimals - you can end up with no decimals, or 4, or en exponential string like "1e-4". To keep older drawings as they are, %g is nevertheless limited if the option to use global decimal is used.

> which is also odd, because "use global decimals" should only affect new documents (it's in italics in the prefs).

The italics is still correct, only for the special case of %g AND the use global decimal option it has an immediate effect. I mean no compromise without any small glitch.

> Also, I got some errors in the report window, like `Warning - no numeric format in formatSpec %.2 - Dimension001`, which seems to imply that something is cutting the format specifier short.

This means that the format specifier is empty:

```
if (formatSpecifier.isEmpty()) {
            Base::Console().Warning("Warning - no numeric format in formatSpec %s - %s\n",
                                    qPrintable(qFormatSpec), getNameInDocument());
            return Base::Tools::toStdString(qFormatSpec);
        }
```

This code was there before my PR as well. So your file has an empty format specifier. But my PR does not change the format specifier, it only reads it. So the question is what emptied the format specifier.

Since this PR is merged, please open a new forum topic and CC me.

**Comment by aapo-aapo** [comment]
> Yes, we agreed on this (you gave me a thumbs up). It is the compromise.

I don't mean that, but the problem is the number of printed decimals changes, if the literal "mm" is printed or not.  For example, for a certain radius, I get either "R13.4 mm" or "R13.3975", if I just click between "Print Units".  Surely that's a bug somewhere?

**Pull request closed without merging**
